plugins/linux-sleep/fu-plugin-linux-sleep.c

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-plugin-vfuncs.h"
+#include "fu-hash.h"
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
+	fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_REQUIRED_FOR, "hsi");
+}
+
+gboolean
+fu_plugin_add_security_attrs (FuPlugin *plugin, GPtrArray *attrs, GError **error)
+{
+	FwupdSecurityAttr *attr;
+	gsize bufsz = 0;
+	g_autofree gchar *buf = NULL;
+	g_autoptr(GFile) file = g_file_new_for_path ("/sys/power/mem_sleep");
+
+	/* load file */
+	if (!g_file_load_contents (file, NULL, &buf, &bufsz, NULL, error)) {
+		g_autofree gchar *fn = g_file_get_path (file);
+		g_prefix_error (error, "could not open %s: ", fn);
+		return FALSE;
+	}
+	attr = fwupd_security_attr_new ("org.kernel.CheckS3Sleep");
+	fwupd_security_attr_set_number (attr, 3);
+	fwupd_security_attr_set_name (attr, "Linux Kernel S3 Sleep");
+	if (g_strstr_len (buf, bufsz, "deep") == NULL) {
+		fwupd_security_attr_add_flag (attr, FWUPD_SECURITY_ATTR_FLAG_SUCCESS);
+	} else {
+		fwupd_security_attr_set_summary (attr, "Deep sleep available");
+	}
```

Comment:
So something I keep going back and forth here; is it necessarily bad that the firmware has s3 methods if the OS isn't using them?  There is a bit in the firmware that gets set setting a preference of whether to use modern standby/s0ix
On Windows this _disables_ S3.  It's not possible to use.
On Linux this sets the default to s2i, but s3 is still possible to invoke with root permissions.

So should we only be focusing on what that bit is set to rather than the complete lack of S3?  I know right now there are a lot of platforms that still offer the S3 ACPI methods but set the bit to S2I.  The S3 methods are there as a contingency in case of power management problems on S2I (which certainly exist).

Reply:
So my idea would be maybe we should instead have this plugin make two assertions.
1. An assertion on the default the firmware is set, this should go into the numerical HSI equation.
2. The "current" value it's set to, this should go into the runtime "!" equation if it's set to S3.

Reply:
If you agree, then the way to evaluate this is by looking at the FADT.  There is an offset that indicates whether to use S2I (At FADT.Flags[21]: https://uefi.org/sites/default/files/resources/Intel_ACPI_Low_Power_S0_Idle.pdf)

The kernel looks at it here to set the default:
https://github.com/torvalds/linux/commit/e870c6c87cf9484090d28f2a68aa29e008960c93

Reply:
> is it necessarily bad that the firmware has s3 methods if the OS isn't using them

I kinda felt the same. 

> On Linux this sets the default to s2i

As in what happens when systemd echos `mem` into `sleep`? Or is it more complicated now.

> by looking at the FADT

How can we do that from userspace?

Reply:
>As in what happens when systemd echos mem into sleep? Or is it more complicated now.

Yes when `mem` is echoed into `sleep` the current looks at the value of `mem_sleep` to determine what to do, s2i or s3.

> How can we do that from userspace?

Same way the DMAR stuff you just did, grab FADT from `/sys/firmware/acpi` and then parse it to figure out that bit in the spec I linked above.

Reply:
Hmm, I don't have a `/sys/firmware/acpi/tables/LPIT` -- do you?

Reply:
Yeah I do have that on my system, but that is not what _needs_ to be parsed.  Just the FADT at that one bit I said above which should be on any system with ACPI 5.0 and later.

Reply:
Ahh, the FADT is `/sys/firmware/acpi/tables/FACP` -- give me 30 mins

Reply:
I've pushed something -- is that what you had in mind? My P50 seems to only support s3 sleep -- if you have a system with LPIT handy I'd appreciate if you could upload to `plugins/acpi-facp/tests/FACP-with-LPIT` or something and add a successful test.

Reply:
Added.

Reply:
Great, thanks. I've rebased this PR on top of FuSecurityAttrs and it's much nicer and easier to understand now. Can you have a look? I can break it up into smaller chunks if that's easier to digest.